### ISSUE TITLE
BB Editor should have a button to insert a percussion track (equivalent to drag-n-drop a sample)

### ISSUE THREAD

**Issue by raindropsfromsky** [LMMS/lmms#4220]
In a new project, the BB Editor starts out with a kicker track, which is not a standard drumkit sound. 
So a composer would want to add them to BB Editor as a common practice.
(In a typical drumkit, there are >7 different sounds.)

But there is no button in the BB Editor to add such sounds directly.
(Drag-n-drop samples from sidebar is not such an elegant method.)
This is a basic deficiency.

**Desired:** 

1. Please provide a button to insert percussion tracks. </br>(equivalent to drag-n-drop of a sample onto BB Editor).
2. It would also be good if the full drumkit is loaded in the BB Editor by default. </br>Then the composer can replace the default sounds with the variants he likes.

**Comment by Spekular** [comment]
I suppose this would be an "Add AFP" button, since this is what plays
samples. With this button you would have to:

Add AFP->Open AFP window->Browse for sample in file browser

This seems like more work than just dragging in a sample to me, but I can't
say it would make anything worse for me.

Adding default samples to the BB editor would be a downgrade imo. I believe
these samples would need to be replaced essentially every time, because
each genre calls for different samples and each producer has their own
sample library and preferences. They would also increase the size of a
default project.

This is a better fit for a custom template. If someone finds that they use
the same samples in the BB editor in many projects, they can create a
template including them. This also lets them personally choose which
samples they want instead of having a "one size fits all" kit built into
LMMS by default.

**Comment by raindropsfromsky** [comment]
No "Add AFP" is how the coder sees it. We should not design a GUI from coders' point of view, but from user's point of view.

When the user adds a BB track, he will need a sound. Thus he will have to browse for it. Therefore the addition of a **Beats** track and identification of a sound for it are two sub-steps. The user will cover both sub-steps in a single workflow. So it will be very natural.

**Comment by raindropsfromsky** [comment]
BTW I agree that each composer may want a different sound for his beats (and basslines).
My focus was on the fact that a single kicker track is just a formality for most composers, and they will have to add new beat tracks. But the BB Editor does not have a button for that. 

It does have a button for the bassline and automation. So why not for beats?

**Comment by tresf** [comment]
I'm not even sure what this is in regards to, to be honest.  DAWs offer `drumkits` but they're similar to ours.  We just lack a decent sampler, so there are no default drumkit presets.  This is a duplicate of https://github.com/LMMS/lmms/issues/1460, which can come with it the presets you're requesting.

But not, there's no such thing as a "drum track" or "beats track".

> No "Add AFP" is how the coder sees it. We should not design a GUI from coders' point of view, but from user's point of view.

Please do not attempt to represent how a "coder" sees things.  Many of us are hobbyist musicians and most of us use several DAWs, these statements are more insulting than they are relevant.

The "AFP" is our instrument.  Ableton has the "Sampler", which we may eventually adopt as well to make the software more friendly, but when you select a drumkit in a commercial DAW, it's generally a superior sampler preset, and it's not offered up as a "track type". I'm not even sure what that is or what is being proposed.

**Comment by Spekular** [comment]
My response has nothing to do with how a coder sees it, it's entirely about
the user facing side.

There is no such thing as a "percussion track". There is a button to add
automation and sample tracks, as you have pointed out. An additional button
would have to add an instrument of some sort, but there are several
instruments. We're left with two options:

1. You want a button that will let you add any of the instruments. This
button couldn't possibly be a single click, and displaying a list of all
instruments available is guaranteed to take up lots of space.

2. You want a button to add a specific instrument. You proposed that your
button would be an alternative to dragging samples from the sidebar, so it
only makes sense that an AFP would be added.

You could have this button immediately open a file browser to select your
sample, I suppose. If this was your intention then I can see this being a
decent alternative to using the sample browser, but I don't believe you
mentioned any details of how the button would work.

**Comment by raindropsfromsky** [comment]
Oops. 
I failed to recognize that "AFP" stood for "Audio File Processor".
(I thought it to be some code block). My apologies.

I am not so familiar with LMMS to recognize these acronyms.

Yes, my suggestion was based on AFP only.

@Spekular I had not thought about having dedicated buttons for different instruments. I guess that's how many Tablet-based apps (e.g. Garage Band) work. My idea was very similar to how SF@ player works. But there, we have to click on the Folder button in the instrument window, which pops up a browser window. In my suggestion, this step should be automatically taken, as follows:

1. User presses the "Add beats track".
2. LMMS ladds a track with 16 steps (assuming meter is 4/4). This track will have AFP.
3. LMMS pops up a browser window so that the user can select the sound sample. </br> The browser window should have a "sound preview" facility so that user can check out the sounds, select what he wants and then press OK. (The sidebar allows us to play the samples. This is equivalent of that.)

@tresf my idea had not extended to presets. I just thought that providing a full drumkit sounds would get the newbies going. (if they can make a decent music with LMMS out of the box, it will make them happier.)

Once they pick up some experience, then they can think of customization, which is easily possible by replacing individual sounds with other options; even without presets.

I am studying up polyphone so that if they want to make a preset (SF2 file), we can guide them.
But that topic is not for this thread! :)

**Comment by tresf** [comment]
You're overthinking "preset".  A drum instrument would have preset drums applied in MIDI drumpad arrangement.

The 16-steps is already a thing.  I still don't understand what's being proposed.

**Comment by raindropsfromsky** [comment]
> You're overthinking "preset". 

I thought you were talking about making of presets inside LMMS. Never mind.

> The 16-steps is already a thing. I still don't understand what's being proposed.

The 16-step part is a bit redundant. My suggestion is to launch a AFP track when user presses **"Add a beats track"** button. After launching AFP, LMMS should automatically execute the next step of browsing for the sample file.

**Comment by tresf** [comment]
> My suggestion is to launch a AFP track when user presses "Add a beats track" button. After launching AFP, LMMS should automatically execute the next step of browsing for the sample file.

This contradicts the bug report I feel.  A beat and the AFP are different concepts.  If you study other DAWs, the beats are generally provided to newcomers as drumkits.  We won't be adding a dedicated button for this, but it would be sensible to add a default track worth the samples one would expect to make drums, perhaps replacing or complimenting the default kicker track.

**Comment by tresf** [comment]
@raindropsfromsky to provide some more context, here's the mockup that @Spekular created for a fork of AudioFileProcessor that would support the idea of a drumkit: https://github.com/LMMS/lmms/pull/4075#issuecomment-354568833

I would like to close this as a duplicate of the general idea of a drum playing plugin.  Whether we adding one to the default template or make a user select a drumkit from the sidebar seems like an easy decision once the plugin is completed.

**Comment by musikBear** [comment]
Drum-tracks in B&B .. confuses me 
To me, it sounds like crossover to Songeditor (?)
I think that some of these consideration has it origin in FLs, where *everything* actually is being *build* in a B&B like feature, and FLs songeditor will simply **paint** these defined sequences -beats as well as melody! -into any number of bars. And that philosophy is ok, but it will be much easier to be 'lacy', and just 'reuse' *same* parts, wheras lmms invites to more creativity, in the various TCOs over n bars ..just saying

>  We just lack a decent sampler

Amen to that! Especially since no vst samplers works.
AFP should hold most of the code needed. Some creative 'reuse' of AFP could be a solution.
However, *my* understanding of a *real* sampler, include the feature to spread *one* sample over n octave, and pitch that to correct length too. Is that a wrong conception. Many seams to regard a plugin as a 'sampler' if it  'just' can have different sampled sounds, on different keys... idk...

**Comment by tresf** [comment]
> spread one sample over n octave, and pitch that to correct length too.

Eventually. 👍

**Comment by raindropsfromsky** [comment]
@tresf  

> A beat and the AFP are different concepts.

How so? I dragged AFP into BB Editor, and it created a 16-step track.
![image](https://user-images.githubusercontent.com/9047168/37075333-60d003fa-21f7-11e8-969e-67c7169dfb7d.png)
This is what I mean by a "beat track".

Now if I load a short-duration sound in AFP, I am done. It can serve as percussion sound.
- AFP can manipulate the sample with this toolbar </br>![image](https://user-images.githubusercontent.com/9047168/37075455-d3286262-21f7-11e8-897c-d6b47e6b2b94.png)
- I can alter the natural pitch of the original sound.

In short, using AFP, I can use any sample as beats.

Now consider the same thing done with a press of a button, instead of dragging and dropping AFP in the BB Editor.

**Comment by raindropsfromsky** [comment]
@musikBear 

> However, my understanding of a real sampler, include the feature to spread one sample over n octave, and pitch that to correct length too. Is that a wrong conception. Many seams to regard a plugin as a 'sampler' if it 'just' can have different sampled sounds, on different keys... idk...

Isn't that what polyphone does? The very idea of SF2 is to take samples of any instrument at some pitches, and then interpolate/extrapolate the same sound at other pitches...

**Comment by tresf** [comment]
> I dragged AFP into BB Editor, and it created a 16-step track.

You're doing it wrong.  Drag a sample in and you don't have to use the plugin explorer.

> How [A beat and the AFP are different concepts] ?

Because the concept of a drumkit requires multiple samples, not just one.  It's superior to your concept and I've cross-linked enhancement requests, WIP code and explained exactly how this will be executed.  Can we let this topic rest please?

**Comment by raindropsfromsky** [comment]
Somewhere in this thread there was a mention of mapping the keys to different drum sounds (drum preset). Such SF2 files are already available. But I did not intend to suggest such a feature.

@tresf 
> If you study other DAWs, the beats are generally provided to newcomers as drumkits.

My second [suggestion ](https://github.com/LMMS/lmms/issues/4220#issue-302569997) was to load one full drumkit by default, so that newbies can start composing without having to look in settings. (I haven't mention this before, but loading standard instruments like Piano, flute and violin in the Song Editor may also help the newbies.) This suggestion is not about core design: Just how the LMMS presents itself to a new user.

> Close this as duplicate of the general idea of a drum playing plugin. 

No this is not about drum-playing plugin, but a different workflow using AFP, as explained [earlier](https://github.com/LMMS/lmms/issues/4220#issuecomment-371029415).

**Comment by tresf** [comment]
> My second suggestion was to load one full drumkit by default, so that newbies can start composing without having to look in settings

Sure.  https://github.com/LMMS/lmms/pull/4075#issuecomment-354568833, #1460, incase you missed them previously.

> a different workflow using AFP, as explained earlier.

No, workflow is identical, you just want a special button, which is inconsistent if it's only for drums.  I do agree that plugins and tracks are obfuscated a bit in the software, but then you're a duplicate of https://github.com/LMMS/lmms/issues/3151.

**Comment by raindropsfromsky** [comment]
@tresf 
Yes, #3151 is exactly what I wanted. Your point is much better.

As regarding loading full drumkit, I just now discovered that the very first template does just that!
It is is a simple matter of using that template while launching a new project.

I think both parts of my suggestions are nicely addressed already.
Closing this issue!
